### PR TITLE
New Fastly terraform module

### PR DESCRIPTION
**Pull request by timetinytim**
Module for creating a Fastly service more easily. Tested this against all existing Fastly infrastructure, and I believe I covered all the use cases for what we have so far. Let me know if this looks OK, and once this is merged I can start putting the rest of the terraform stuff in place.

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+variable "as_blocklist_name" {
+  description = "Name of the AS blocklist"
+  type        = string
+  default     = "Blocked AS Numbers"
+}
+
+variable "as_request_blocklist" {
+  description = "List of Autonomous Systems (AS) to block from making explicit requests."
```
Comment (renchap) [inline variables.tf:229]: What is an explicit request here?
We may want to define it, or make the name cleared

File: variables.tf [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+variable "activate" {
+  description = "Conditionally prevents the Service from being activated. The apply step will continue to create a new draft version but will not activate it if this is set to false."
+  type        = bool
+  default     = true
+}
+
+variable "default_ttl" {
+  description = "The default Time-to-live (TTL) for requests."
+  type        = number
+  default     = 3600
```
Comment (renchap) [inline variables.tf:25]: Isnt 3600s too high as a default TTL?

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    method            = optional(string, "GET")
+    threshold         = optional(number, 1)
+    timeout           = optional(number, 5000)
+    window            = optional(number, 2)
+  }))
+  default = []
+}
+
+# Custom headers
+variable "headers" {
```
Comment (renchap) [inline variables.tf:104]: What is this used for?

File: variables.tf [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+variable "backend_ssl_ca_cert" {
+  description = "CA certificate attached to origin. Used to pull value from Terraform Cloud."
+  type        = string
+  sensitive   = true
+  default     = ""
+}
+
+# Healthcheck endpoints.
+variable "healthchecks" {
+  description = "Endpoints to periodically test for backend health."
```
Comment (renchap) [inline variables.tf:84]: Maybe we could have a default value here, as all Mastodon instances should use the same path for health checks?

**Review by renchap — COMMENTED** [review]
Some general questions:
- Should we have some default VCL snippets, for example the ones setting correct cache headers for some queries?
- Should the AS & IP access lists be all managed by Terraform? This is probably best, but when under DDOS you may want to very quicly block some things without taking the time to commit + apply the change. I guess its fine, but it will need to be reflected into the TF variables once done on-the-fly?

**Comment by renchap** [comment]
Do you have an example of how you are calling this module to match the currently deployed config? It would help ensuring everything looks good!

Also, should we also create a second service for the media files in the same module, or as a separate modules? Those domains have some specific config as well

File: variables.tf (hunk 1 above)
Comment (timetinytim) [reply variables.tf:229]: Good point. Let me tweak it a bit.

File: variables.tf (hunk 3 above)
Comment (timetinytim) [reply variables.tf:104]: This is equivalent to the [Headers configuration](https://docs.fastly.com/en/guides/adding-or-modifying-headers-on-http-requests-and-responses) that you can add to services. There are a couple services that have custom headers specified (for example, `http.Fastly-Purge-Requires-Auth` for the `files-*` services)

**Comment by timetinytim** [comment]
I pushed my current application of this module to this branch here. Every service in Fastly is captured here.
https://github.com/mastodon/terraform/tree/INF-94-fastly-terraform

I considered a second module for the files services, but honestly the only differences are omissions of things, rather than additions, so I figured one module was enough.

We could definitely do two modules though, in which case we could set more defaults and make the actual infra terraform files smaller. I will think about this~

**Comment by timetinytim** [comment]
Renaud had some really good suggestions for how to better structure this module: this should be more specific to our own use cases (i.e. a module centered around specifically creating the services for `mastodon.online` and `mastodon.social`. So I will close this PR, restructure it, and re-open another PR once that is finished.

**Pull request closed without merging**